consensus/core/src/linearizer.rs

```diff
@@ -0,0 +1,457 @@
+    ) -> Self {
+        Self {
+            leader,
+            blocks,
+            timestamp_ms,
+            commit_index,
+        }
+    }
+
+    pub fn new_from_commit_data(commit_data: Commit, block_store: Arc<dyn Store>) -> Self {
+        let mut leader_block_idx = None;
+        let commit_blocks = block_store
+            .read_blocks(&commit_data.blocks)
+            .expect("We should have the block referenced in the commit data");
+        let blocks = commit_blocks
+            .into_iter()
+            .enumerate()
+            .map(|(idx, commit_block_opt)| {
+                let commit_block = commit_block_opt
+                    .expect("We should have the block referenced in the commit data");
+                if commit_block.reference() == commit_data.leader {
+                    leader_block_idx = Some(idx);
+                }
+                commit_block
+            })
+            .collect::<Vec<_>>();
+        let leader_block_idx = leader_block_idx.expect("Leader block must be in the sub-dag");
+        let leader_block_ref = blocks[leader_block_idx].reference();
+        let timestamp_ms = blocks[leader_block_idx].timestamp_ms();
+        CommittedSubDag::new(leader_block_ref, blocks, timestamp_ms, commit_data.index)
+    }
+
+    /// Sort the blocks of the sub-dag by round number. Any deterministic algorithm works.
+    pub fn sort(&mut self) {
+        self.blocks.sort_by_key(|x| x.round());
+    }
+}
+
+impl Display for CommittedSubDag {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "CommittedSubDag(leader={}, index={}, blocks=[",
+            self.leader.digest, self.commit_index
+        )?;
+        for (idx, block) in self.blocks.iter().enumerate() {
+            if idx > 0 {
+                write!(f, ", ")?;
+            }
+            write!(f, "{}", block.digest())?;
+        }
+        write!(f, "])")
+    }
+}
+
+impl fmt::Debug for CommittedSubDag {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}@{}(", self.leader, self.commit_index)?;
+        for block in &self.blocks {
+            write!(f, "{}, ", block.reference())?;
+        }
+        write!(f, ")")
+    }
+}
+
+/// Expand a committed sequence of leader into a sequence of sub-dags.
+#[allow(unused)]
+pub struct Linearizer {
+    // Persistent storage for committed blocks.
+    block_store: Arc<dyn Store>,
+    /// Keep track of all committed blocks to avoid committing the same block twice.
+    committed: HashSet<BlockRef>,
+    /// Keep track of the index of last linearized commit
+    last_commit_index: u64,
+}
+
+#[allow(unused)]
+impl Linearizer {
+    pub fn new(block_store: Arc<dyn Store>) -> Self {
+        Self {
+            block_store,
+            committed: Default::default(),
+            last_commit_index: Default::default(),
+        }
+    }
+
+    pub fn recover_state(&mut self, recovered_state: &CommitObserverRecoveredState) {
```

Comment:
I believe for security reasons we might don't want this method to be called more than once, so maybe we could offer the `recovered_state` as part of the `new` method or have a separate one which creates the `Linearizer` like `from_state(recovered_state: &CommitObserverRecoveredState)`.

My understanding is that `CommitObserverRecoveredState` will be shared between the commit observer and the linearizer, but are we certain that the commit observer needs to have any state recovered? If not then we might just leave the linearizer recover on its own by utilising/query the passed in `block_store` since we already inject it

Reply:
Btw recovering all the committed block refs seems to be, at least in theory and we can confirm that, a quite expensive recovery approach and we might need to think of alternatives here. Looking on the `collect_sub_dag` method I believe switching to the Narwhal like approach where we are checking the highest committed round for an authority and not allow to commit anything bellow might be less expensive in terms of memory.

Reply:
Removed the recovery mechanism in linearizer and will rely on `DagState` to maintain the last commit state correctly. Also switched to using the `last_committed_rounds` approach as we did in NW.